### ISSUE TITLE
jsonpath: missing `left' in filter has confusing error message and fails unexpectedly

### ISSUE THREAD

**Issue by stevekuznetsov** [kubernetes/kubernetes#45546]
On [current master](https://github.com/kubernetes/kubernetes/blob/a7fa2aef8deb40eb1d159dbf7f0d8ea7d6d10dac/staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go#L458-L460), `jsonpath` will fail to evaluate a filter when the `'left'` member does not exist:

```go
if len(lefts) != 1 {
    return input, fmt.Errorf("can only compare one element at a time")
}
```

For instance, if I were to execute

```
kubectl get pods -o jsonpath='{.items[@.metadata.annotations.foo=="bar"].metadata.name}'
```

Where not all `pods` in that namespace had the `foo` annotation, this would fail with the error message:

```
error: error executing jsonpath "{.items[?(@.metadata.annotations.foo==\"bar\")].metadata.name}": can only compare one element at a time
```

This is confusing as it leads me to think more than one item was selected, so while it is not technically incorrect the error message is misleading.

I would furthermore expect `jsonpath` to more gracefully handle this case -- a missing `'left'` here means that the filter should fail, not error.

**Comment by mengqiy** [comment]
> a missing 'left' here means that the filter should fail, not error.

What do you mean by `fail`? How is it different from `error`?

**Comment by stevekuznetsov** [comment]
As in, don't add that item to the list of items on which the following access is made. When I do:

```
kubectl get pod mypod -o jsonpath='{.metadata.annotations.foo}'
```

And there is no `foo` annotation on `mypod`, I just get nothing back -- no error, just nothing. Similarly, i would have expected the filtered expression to give me the `.metadata.name` for all of the pods that have an annotation `foo` with value `bar`. That, anyway is what I understand the JSONPath syntax to mean. When you write

```
kubectl get pods -o jsonpath='{.items[@.metadata.annotations.foo=="bar"].metadata.name}'
```

You are saying:

> Give me the name of all pods that have annotation `foo` with value `bar`.

**Comment by mengqiy** [comment]
cc @smarterclayton @ncdc since you authored https://github.com/kubernetes/kubernetes/pull/31714 and https://github.com/kubernetes/kubernetes/pull/39486.
I'm not sure how we want to fix this.

**Comment by ncdc** [comment]
Taking a look now

**Comment by ncdc** [comment]
@stevekuznetsov @mengqiy it looks like you need to do this (note that the annotations portion is wrapped by `?( )`):

```
kubectl get pods -o jsonpath='{.items[?(@.metadata.annotations.foo=="bar")].metadata.name}'
```

I wrote a unit test and it worked as expected. PTAL!

**Comment by stevekuznetsov** [comment]
I'm pretty sure that was a typo on my end but let me check -- do you have a reproducer script?

**Comment by stevekuznetsov** [comment]
I see an entirely different error today... weird:

```
kind: Service
apiVersion: v1
metadata:
  name: my-service
  annotations:
    foo: bar
spec:
  selector:
    app: MyApp
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
```

```
kind: Service
apiVersion: v1
metadata:
  name: my-other-service
spec:
  selector:
    app: MyApp
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
```
$ kubectl create -f /tmp/my-service.yml
service "my-service" created
$ kubectl create -f /tmp/my-other-service.yml
service "my-other-service" created
$ kubectl get services -o jsonpath='{ .items[?(@.metadata.annotations.foo=="bar")].metadata.name }'
Error executing template: can only compare one element at a time. Printing more information for debugging the template:
	template was:
		{ .items[?(@.metadata.annotations.foo=="bar")].metadata.name }
	object given to jsonpath engine was:
		map[string]interface {}{"kind":"List", "apiVersion":"v1", "metadata":map[string]interface {}{}, "selfLink":"", "resourceVersion":"", "items":[]interface {}{map[string]interface {}{"kind":"Service", "apiVersion":"v1", "metadata":map[string]interface {}{"namespace":"myproject", "selfLink":"/api/v1/namespaces/myproject/services/my-other-service", "uid":"45824848-5690-11e7-b13c-54ee753e2644", "resourceVersion":"1323", "creationTimestamp":"2017-06-21T14:45:29Z", "name":"my-other-service"}, "spec":map[string]interface {}{"clusterIP":"172.30.190.175", "type":"ClusterIP", "sessionAffinity":"None", "ports":[]interface {}{map[string]interface {}{"protocol":"TCP", "port":80, "targetPort":9376}}, "selector":map[string]interface {}{"app":"MyApp"}}, "status":map[string]interface {}{"loadBalancer":map[string]interface {}{}}}, map[string]interface {}{"status":map[string]interface {}{"loadBalancer":map[string]interface {}{}}, "kind":"Service", "apiVersion":"v1", "metadata":map[string]interface {}{"name":"my-service", "namespace":"myproject", "selfLink":"/api/v1/namespaces/myproject/services/my-service", "uid":"3d493db8-5690-11e7-b13c-54ee753e2644", "resourceVersion":"1317", "creationTimestamp":"2017-06-21T14:45:16Z", "annotations":map[string]interface {}{"foo":"bar"}}, "spec":map[string]interface {}{"ports":[]interface {}{map[string]interface {}{"port":80, "targetPort":9376, "protocol":"TCP"}}, "selector":map[string]interface {}{"app":"MyApp"}, "clusterIP":"172.30.216.228", "type":"ClusterIP", "sessionAffinity":"None"}}}}

error: error executing jsonpath "{ .items[?(@.metadata.annotations.foo==\"bar\")].metadata.name }": can only compare one element at a time
```

**Comment by ncdc** [comment]
Ok so you have 1 object with annotations, and 1 without. Let me look into that case.

**Comment by stevekuznetsov** [comment]
Let me add another annotation to `my-other-service`

**Comment by stevekuznetsov** [comment]
No difference if both have annotations:

```
$ kubectl get service my-service -o jsonpath={.metadata.annotations}
map[foo:bar]
$ kubectl annotate service my-other-service baz=far
service "my-other-service" annotated
$ kubectl get service my-other-service -o jsonpath={.metadata.annotations}
map[baz:far]
$ kubectl get services -o jsonpath='{ .items[?(@.metadata.annotations.foo=="bar")].metadata.name }'
Error executing template: can only compare one element at a time. Printing more information for debugging the template:
	template was:
		{ .items[?(@.metadata.annotations.foo=="bar")].metadata.name }
	object given to jsonpath engine was:
		map[string]interface {}{"kind":"List", "apiVersion":"v1", "metadata":map[string]interface {}{}, "selfLink":"", "resourceVersion":"", "items":[]interface {}{map[string]interface {}{"metadata":map[string]interface {}{"uid":"45824848-5690-11e7-b13c-54ee753e2644", "resourceVersion":"1384", "creationTimestamp":"2017-06-21T14:45:29Z", "annotations":map[string]interface {}{"baz":"far"}, "name":"my-other-service", "namespace":"myproject", "selfLink":"/api/v1/namespaces/myproject/services/my-other-service"}, "spec":map[string]interface {}{"type":"ClusterIP", "sessionAffinity":"None", "ports":[]interface {}{map[string]interface {}{"protocol":"TCP", "port":80, "targetPort":9376}}, "selector":map[string]interface {}{"app":"MyApp"}, "clusterIP":"172.30.190.175"}, "status":map[string]interface {}{"loadBalancer":map[string]interface {}{}}, "kind":"Service", "apiVersion":"v1"}, map[string]interface {}{"metadata":map[string]interface {}{"name":"my-service", "namespace":"myproject", "selfLink":"/api/v1/namespaces/myproject/services/my-service", "uid":"3d493db8-5690-11e7-b13c-54ee753e2644", "resourceVersion":"1317", "creationTimestamp":"2017-06-21T14:45:16Z", "annotations":map[string]interface {}{"foo":"bar"}}, "spec":map[string]interface {}{"ports":[]interface {}{map[string]interface {}{"protocol":"TCP", "port":80, "targetPort":9376}}, "selector":map[string]interface {}{"app":"MyApp"}, "clusterIP":"172.30.216.228", "type":"ClusterIP", "sessionAffinity":"None"}, "status":map[string]interface {}{"loadBalancer":map[string]interface {}{}}, "kind":"Service", "apiVersion":"v1"}}}

error: error executing jsonpath "{ .items[?(@.metadata.annotations.foo==\"bar\")].metadata.name }": can only compare one element at a time
```

**Comment by ncdc** [comment]
Ok, I'll have to test using kubectl

**Comment by stevekuznetsov** [comment]
I'm the Origin bundled `kubectl` but I don't think there are any patches on JSONPath there:

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1+5115d708d7", GitCommit:"010d313", GitTreeState:"clean", BuildDate:"2017-06-12T20:41:29Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1+5115d708d7", GitCommit:"010d313", GitTreeState:"clean", BuildDate:"2017-06-07T23:14:34Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by ncdc** [comment]
I'm able to reproduce in the unit test when 1 object has annotations and the other does not, if allowMissingKeys is true for the jsonpath parser.

**Comment by stevekuznetsov** [comment]
I see. Let me know if I can be of more help.

**Comment by ncdc** [comment]
I think I have a fix. PR coming soon.

**Comment by ncdc** [comment]
Possible fix: #47846
